### PR TITLE
add marketplace capability

### PR DESCRIPTION
**Pull request by wills2133**
For more information about how to contribute to this repo, visit [this page](https://github.com/OfficeDev/microsoft-teams-library-js/blob/main/CONTRIBUTING.md).

## Description

To enable Teams users a more flexible purchasing experience, we introduce the marketplace capability to make checkout flow can be powered by a Teams app. A Teams app can leverage the marketplace capability to get and update a cart owned by the Teams client.

### Main changes in the PR:

1. add `marketplace` capability.
2. add `getCart`, `addOrUpdateCartItems`, `removeCartItems`, and `updateCartStatus` to interface with `cart` in the host.

### Validation performed:

1. Verify  `getCart`, `addOrUpdateCartItems`, `removeCartItems`, and `updateCartStatus` can modify the `cart` object in the MetaOS host.

### Unit Tests added:

Yes

### End-to-end tests added:

Yes

### Change file added:

Yes

### Related PRs:

[adding marketplace capability in hub sdk](https://office.visualstudio.com/ISS/_git/metaos-hub-sdk/pullrequest/2001855)

### CODE REVIEW

**Comment by wills2133** [comment]
> @wills2133 please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@microsoft-github-policy-service agree company="Microsoft"

File: packages/teams-js/src/public/interfaces.ts [hunk 1]
```diff
@@ -1053,3 +1056,95 @@ export interface AdaptiveCardVersion {
   /** Represents the minor version number. */
   minorVersion: number;
 }
+
+/**
+ * the cart object for app checkout flow
```
Comment (karthikeyan-sh) [inline packages/teams-js/src/public/interfaces.ts:1061]: will it be ok to annotate with @beta here initially

File: packages/teams-js/src/public/interfaces.ts [hunk 2]
```diff
@@ -1053,3 +1056,95 @@ export interface AdaptiveCardVersion {
...
+  /**
+   * indicate checking out is completed and
+   * response a new cart in the next getCart call
+   */
+  Processed = 'Processed',
+  /**
+   * indicate checking out if fail and
+   * response a new cart in the next getCart call
+   */
+  Error = 'Processed',
```
Comment (karthikeyan-sh) [inline packages/teams-js/src/public/interfaces.ts:1141]: typo here

File: packages/teams-js/src/public/interfaces.ts [hunk 3]
```diff
@@ -1053,3 +1056,95 @@ export interface AdaptiveCardVersion {
...
+  /** the cart is created by admin of an organization. */
+  Admin = 'Admin',
+  /** the cart is created by end user of an organization. */
+  IW = 'IW',
+}
+
+/** the status of the cart. */
+export enum CartStatus {
+  /** cart is created but not checked out yet.*/
+  Opened = 'Opened',
```
Comment (karthikeyan-sh) [inline packages/teams-js/src/public/interfaces.ts:1129]: could this just be open?

File: packages/teams-js/src/public/interfaces.ts [hunk 4]
```diff
@@ -1053,3 +1056,95 @@ export interface AdaptiveCardVersion {
...
+  /** Represents the major version number. */
+  majorVersion: number;
+  /** Represents the minor version number. */
+  minorVersion: number;
+}
+
+/** the cart info. */
+interface CartInfo {
+  /** the country market that where the products are . */
+  readonly market: Country;
```
Comment (karthikeyan-sh) [inline packages/teams-js/src/public/interfaces.ts:1087]: market: Market (instead of country?)

File: packages/teams-js/src/public/interfaces.ts [hunk 5]
```diff
@@ -1053,3 +1056,95 @@ export interface AdaptiveCardVersion {
...
+  /** currency code for the cart item price. */
+  readonly currency: Currency;
+}
+
+/** the cart items. */
+interface Item {
+  /** the id of the cart item. */
+  readonly id: string;
+  /** the display name of the cart item. */
+  name: string;
```
Comment (karthikeyan-sh) [inline packages/teams-js/src/public/interfaces.ts:1103]: is this is also read only? if id is readonly

File: packages/teams-js/src/public/interfaces.ts [hunk 6]
```diff
@@ -1053,3 +1056,95 @@ export interface AdaptiveCardVersion {
...
+ */
+export interface Cart {
+  /** version of the cart. */
+  readonly version: CartVersion;
+  /** the id of the cart. */
+  readonly id: string;
+  /** the cart info. */
+  cartInfo: CartInfo;
+  /** the cart items. */
+  cartItems: CartItem[];
```
Comment (karthikeyan-sh) [inline packages/teams-js/src/public/interfaces.ts:1071]: trace id?

File: packages/teams-js/src/public/interfaces.ts (hunk 4 above)
Comment (wills2133) [reply packages/teams-js/src/public/interfaces.ts]: update schema in marketplace.ts

File: packages/teams-js/src/public/interfaces.ts (hunk 2 above)
Comment (wills2133) [reply packages/teams-js/src/public/interfaces.ts]: fixed in marketplace.ts

File: packages/teams-js/src/public/interfaces.ts (hunk 3 above)
Comment (wills2133) [reply packages/teams-js/src/public/interfaces.ts]: updated in marketplace.ts

File: packages/teams-js/src/public/marketplace.ts [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+ */
+interface Item {
+  /** The id of the cart item. */
+  readonly id: string;
+  /** The display name of the cart item. */
+  name: string;
+  /** The quantity of the cart item. */
+  quantity: number;
+  /** The price of the single cart item. */
+  price: number;
```
Comment (karthikeyan-sh) [inline packages/teams-js/src/public/marketplace.ts:66]: is number a decimal?

File: packages/teams-js/src/public/marketplace.ts [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+}
+
+/**
+ * Represents the cart information
+ *
+ * @beta
+ */
+interface CartInfo {
+  /** The country market where the products are selling. */
+  readonly market: Market;
```
Comment (karthikeyan-sh) [inline packages/teams-js/src/public/marketplace.ts:42]: can this be a optional string?

File: packages/teams-js/src/public/marketplace.ts [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+  /** The identifier to tell the cart is checked out by admin or information worker */
+  readonly intent: Intent;
+  /** Locale for the user */
+  readonly locale: Locale;
```
Comment (karthikeyan-sh) [inline packages/teams-js/src/public/marketplace.ts:46]: can this be a optional string?

File: packages/teams-js/src/public/marketplace.ts [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+/**
+ * Represents the cart information
+ *
+ * @beta
+ */
+interface CartInfo {
+  /** The country market where the products are selling. */
+  readonly market: Market;
+  /** The identifier to tell the cart is checked out by admin or information worker */
+  readonly intent: Intent;
```
Comment (karthikeyan-sh) [inline packages/teams-js/src/public/marketplace.ts:44]: can this be a optional string?

File: packages/teams-js/src/public/marketplace.ts [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+}
+
+/**
+ * Represents the persona creating the cart.
+ *
+ * @beta
+ */
+export enum Intent {
+  /** The cart is created by admin of an organization. */
+  AdminUser = 'AdminUser',
```
Comment (karthikeyan-sh) [inline packages/teams-js/src/public/marketplace.ts:88]: can this be 
AdminTAC
AdminClient
IWClient 

and this enum will grow, how easy is it to add more to this enum without making much changes to the public interface?
Would it be better to keep it a string?

File: packages/teams-js/src/public/marketplace.ts [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+/**
+ * Represents the status of the cart.
+ *
+ * @beta
+ */
+export enum CartStatus {
+  /** Cart is created but not checked out yet. */
+  Open = 'Open',
+  /** Cart is checked out but not completed yet. */
+  Processing = 'Processing',
```
Comment (karthikeyan-sh) [inline packages/teams-js/src/public/marketplace.ts:102]: while implementing logic we will use open and processed / will avoid processing state on our end.

File: packages/teams-js/src/public/marketplace.ts [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+/**
+ * Represents the parameters to update the cart status.
+ *
+ * @beta
+ */
+export interface UpdateCartStatusParams {
+  /** Status of the cart. */
+  cartStatus: CartStatus;
+  /** Extra info to the status. */
+  message?: string;
```
Comment (karthikeyan-sh) [inline packages/teams-js/src/public/marketplace.ts:123]: what can this message carry?

**Comment by TrevorJoelHarris** [comment]
Closing this, you opened a newer version [here](https://github.com/OfficeDev/microsoft-teams-library-js/pull/1802)

**Pull request closed without merging**